### DISCUSSION TITLE
Raw string literals in nushell

### DISCUSSION THREAD

**Discussion by nobiit** [nushell/nushell#10552] [Q&A]
I need to store a safe string, which can have any characters. In bash there is multi-line string, is there any way in nushell?

**Comment by amtoine** [comment]
you can just do
```nushell
"my
multi
line
string"
```
and it will span 4 lines :yum:

**Reply by nobiit** [reply]
In that string containing special characters for example `\a`, I want the actual value to be saved as `\a` without changing it to `\\a`

**Reply by ghost** [reply]
you can use single-quotes:
```nu
let a = 'foo\n\a\\
\ "
'
```
but if it contains `'` it wont work (theres probably a workaround, but i dont know/need it)

**Comment by ghost** [comment]
2 methods, which are especially a bit nicer for reading within scripts than amtoine's answer:
```nu
some_already_existent_indentation {
  let example = (
    # the let (and above) is just for some example context
    # the `(` and `)` prevent early parsing and allow us to make this multi line
    "first line\n" +
    # the `\n` gets transformed into a newline-character (see escape codes)
    # the `+` adds the next string onto the end of the first one
    "second line\n" +
    "third line"
  )
}
```

```nu
some_context {
  let foo = ([  # create a list of strings
    "first line"  # no " \n" here (which reduces git-diffs, human error, work in really long texts, etc)
    "second line"
    "third line"
  ] | str join "\n")  # tape the strings in the list together using newline characters
}
```
both are also adaptable for some other scenarios (but i prefer to use the second one in most of those).
examples:
```nu
ls | each {|i|
  $"($i.name) is ($i.size) big."
} | str join "\n"

mut a = ["first line"]
$a = ($a | append "second line")
for i in 1..10 {
  $a = ($a | append $"($n)th line")
}
print ($a | str join "\n")

mut a = "first line"
$a = ($a + "\nsecond line")
for n in 1..10 {
  $a = ($a + $"\n($n)th line")
}
print $a
```

**Reply by nobiit** [reply]
I need to assign strings in one go (to quickly copy/paste from another place, for example), so logic, arrays, etc. are not recommended.

**Reply by murlakatamenka** [reply]
> ```nu
> some_already_existent_indentation {
>   let example = (
>     # the let (and above) is just for some example context
>     # the `(` and `)` prevent early parsing and allow us to make this multi line
>     "first line\n" +
>     # the `\n` gets transformed into a newline-character (see escape codes)
>     # the `+` adds the next string onto the end of the first one
>     "second line\n" +
>     "third line"
>   )
> }
> ```

This is the best solution I've found so far, thanks for sharing.

Also it's almost like Python:

```py
s = (
    f"name: {person.name}\n"  # add some ansi colors for a very long string
    f"age: {person.age}"      # here too
)

print(s)

```

**Comment by amtoine** [comment]
@nobiit 
for now, Nushell does not have a multiline string that does not escape, as `r#"..."#` in Rust for instance :thinking:

**Reply by nobiit** [reply]
Yep, I like this too. Now we have to wait?

**Reply by amtoine** [reply]
not sure if we ever took a decision on that matter :thinking:

**Reply by weirdan** [reply]
raw strings are now a thing: https://www.nushell.sh/book/working_with_strings.html#raw-strings

**Comment by lavafroth** [comment]
I stumbled upon this issue when implementing #10600.
Consider the contrived example of a file with the name
```
`"'file
```

This path can neither be completed nor represented in Nu.

**Comment by fdncred** [comment]
I still think we should have raw string literals similar to rust like `r'some raw string'`. I create a PR for it a while back https://github.com/nushell/nushell/pull/9956
